File: leetcode/src/lru/LRUCache.java
package lru;

import java.util.HashMap;

public class LRUCache {

    private HashMap<Integer, Node> map;
    private DoubleList cache;
    private int cap;

    public LRUCache(int capacity) {
        map = new HashMap<>(capacity);
        cache = new DoubleList();
        cap = capacity;
    }

    public int get(int key) {
        if (!map.containsKey(key)) return -1;
        int val = map.get(key).val;
        put(key, val);
        return val;
    }

    public void put(int key, int value) {
        Node n = new Node(key, value);
        if (map.containsKey(key)) { //已经存在
            cache.remove(map.get(key)); //移除原来节点
            cache.addFirst(n); //将节点重新插到头部
            map.put(key, n);//更新 map
        } else {
            if (cache.size() == cap) { //容量已满
                Node last = cache.removeLast(); //删除最后节点
                map.remove(last.key); //同时删除 map 中的 value
            }
            cache.addFirst(n); //插入头节点
            map.put(key, n); //更新map
        }
    }
}

File: leetcode/src/lru/LruCache.kt
package lru

class LruCache(val capacity: Int) {

    var map: HashMap<Int, DEntryNode> = HashMap()
    var list: DoubleLinkList = DoubleLinkList()

    fun put(key: Int, value: Int) {
        val n = DEntryNode(key, value)
        if (map.containsKey(key)) {
            val exist = map.get(key)
            list.remove(exist)
            list.addFirst(n)
            map.put(key, n)
        } else {
            if (list.size == capacity) {
                val last = list.removeLast()
                map.remove(last?.key)
            }
            list.addFirst(DEntryNode(key, value))
            map.put(key, n)
        }
    }

    fun get(key: Int): Int {
        if (!map.containsKey(key)) return -1
        val value = map.get(key)?.value
        put(key, value!!)
        return value
    }
}

fun main(args: Array<String>) {
    val cache = LruCache(5)
    cache.put(1,1)
    cache.put(2,2)
    cache.put(3,3)
    cache.put(4,4)
    cache.put(5,5)
    cache.put(6,6)
    println(cache.get(1))
    println(cache.get(2))
}



